### PR TITLE
Add back bound_elemwise. 

### PR DESCRIPTION
**Pull request by twiecki**
Second attempt at #1579. Misses tests.

Previous: #1591

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -146,8 +146,8 @@ def random(self, point=None, size=None, repeat=None):
     def logp(self, value):
         lower = self.lower
         upper = self.upper
-        return bound(-tt.log(upper - lower),
-                     value >= lower, value <= upper)
+        return bound_elemwise(-tt.log(upper - lower),
```
Comment (fonnesbeck) [inline pymc3/distributions/continuous.py]: shouldn't this require `ones_like` also, if we are going that route to deal with the broadcasting issue?

Comment (twiecki) [reply pymc3/distributions/continuous.py]: `bound_elemwise` I think is supposed to handle that case by broadcasting via alltrue_elemwise.

**Comment by jsalvatier** [comment]
Almost all the current calls to `bound`, especially in continuous.py and discrete.py, should be replaced with `bound_elemwise`. 

I'm somewhat worried that this will break other people's code. We could instead have `bound` and `bound_multivariate`, but I feel like that's going to lead to more future coding errors. For example if, someone looks at an elemwise distribution as a template for their multivariate distribution (the reverse seems less likely).

Thoughts?

Otherwise looks good to me.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (fonnesbeck) [reply pymc3/distributions/continuous.py]: Does it not do that for `DiscreteUniform` too?

File: pymc3/distributions/discrete.py [hunk 2]
```diff
@@ -348,8 +348,8 @@ def random(self, point=None, size=None, repeat=None):
     def logp(self, value):
         upper = self.upper
         lower = self.lower
-        return bound(-tt.log(upper - lower + 1),
-                     lower <= value, value <= upper)
+        return bound_elemwise(-tt.log(upper - lower + 1) * tt.ones_like(value),
```
Comment (jsalvatier) [inline pymc3/distributions/discrete.py]: Like Chris says, seems like this `ones_like` shouldn't be necessary?

Comment (twiecki) [reply pymc3/distributions/discrete.py]: Ah yes, that's a wrong commit.

**Comment by fonnesbeck** [comment]
@jsalvatier can we have `bound_elemwise` infer a multivariate from the shape of the inputs? Or use a flag?

**Comment by twiecki** [comment]
Some more progress.

**Comment by jsalvatier** [comment]
@fonnesbeck, a flag would be fine with me.

Unless I'm missing something, I don't think we can infer it since you could always have a multivariate distribution that has the same shape of parameters for an elemwise distribution.

**Comment by twiecki** [comment]
Opinions on names for the flag? `broadcast=True`?

**Comment by fonnesbeck** [comment]
I'd name it `FlaggyMcFlagface=True`, but that's just me.

@jsalvatier I was thinking in terms of the values being 2D in multivariate distributions. A flag is probably better, though, in any case.

**Comment by jsalvatier** [comment]
If a flag, I would choose `broadcast_conditions` or `multivariate` (with a good description in the doc). `broadcast` seems too generic.

@fonnesbeck sure, that's the most common case, but it really has to cover all the cases well or it will lead to problems.

**Comment by springcoil** [comment]
So I'm fine with this but what edge cases are we not covering.

**Comment by twiecki** [comment]
@jsalvatier OK, I added a `broadcast_conditions` kwarg.

**Pull request merged**